### PR TITLE
Implement `-c #none`

### LINKED ISSUE #1294: Allow a color in an inline palette spec to be `none`
This would make it more intuitive to specify the 0/1/2/3 slot placement of fewer than 4 colors, e.g.:

```
$ rgbgfx -c '#none,#f00,#none,#00f` -o silhouette.2bpp silhouette.png
```

(We need the no-color placeholder to start with a `#` in case the first slot uses it, because the whole thing has to start with a hash for the parser to identify it.)

### PR DESCRIPTION
**Pull request by eievui5**
Closes #1294

### CODE REVIEW

File: man/rgbgfx.1 [hunk 1]
```diff
@@ -124,12 +124,14 @@ If
 begins with a hash character
 .Ql # ,
 it is treated as an inline palette specification.
-It should contain a comma-separated list of hexadecimal colors, each beginning with a hash.
-Colors in are accepted either as
+It should contain a comma-separated list of hexadecimal colors (or the word "none"), each beginning with a hash.
```
Comment (Rangi42) [inline man/rgbgfx.1:127]: Trying to make it very clear that the placeholder is `#none`, not `none`.
```suggestion
It should contain a comma-separated list of hexadecimal colors, each beginning with a hash, or
.Ql #none
to leave a gap in the palette.
```

File: man/rgbgfx.1 [hunk 2]
```diff
@@ -124,12 +124,14 @@ If
...
+Colors are accepted either as
 .Ql #rgb
 or
 .Ql #rrggbb
 format.
+To leave a gap in the palette, write
+.Ql #none .
```
Comment (Rangi42) [inline man/rgbgfx.1:134]: Continued from above:
```suggestion
```
Alternatively, maybe leave this explicit explanation of `#none` here, and don't even mention it in the above part.

File: src/gfx/main.cpp [hunk 3]
```diff
@@ -761,9 +761,16 @@ int main(int argc, char *argv[]) {
 		}());
 		if (options.palSpecType == Options::EXPLICIT) {
 			fputs("\t[\n", stderr);
-			for (std::array<Rgba, 4> const &pal : options.palSpec) {
-				fprintf(stderr, "\t\t#%06x, #%06x, #%06x, #%06x,\n", pal[0].toCSS() >> 8,
-				        pal[1].toCSS() >> 8, pal[2].toCSS() >> 8, pal[3].toCSS() >> 8);
+			for (std::array<std::optional<Rgba>, 4> const &pal : options.palSpec) {
```
Comment (Rangi42) [inline src/gfx/main.cpp:764]: This type is now long enough to justify using `auto`.
```suggestion
			for (const auto &pal : options.palSpec) {
```

**Review by Rangi42 — CHANGES_REQUESTED** [review]
Documentation suggestions

File: src/gfx/main.cpp [hunk 4]
```diff
@@ -761,9 +761,16 @@ int main(int argc, char *argv[]) {
...
+				fputs("\t\t", stderr);
+				for (auto& color : pal) {
```
Comment (Rangi42) [inline src/gfx/main.cpp:766]: Formatting convention.
```suggestion
				for (auto &color : pal) {
```

File: man/rgbgfx.1 (hunk 2 above)
Comment (ISSOtm) [reply man/rgbgfx.1:134]: Then that would be

```suggestion
To leave a gap in the palette,
.Ql #none
can be used instead of one or more colours.
```

Comment (pinobatch) [reply man/rgbgfx.1:134]: Would it be worth moving the thing about quoting this argument up a bit? Right now it implies that only semicolons have to be quoted, but it should really apply to everything to avoid `#` getting parsed as a line comment.

File: src/gfx/main.cpp [hunk 5]
```diff
@@ -840,18 +847,42 @@ auto Palette::begin() -> decltype(colors)::iterator {
...
+		--it; // The array is not empty, so it's fine to decrement pre-emptively.
+		// If the iterator points at a non-empty element, then it is the end
+		// (being followed only by empty slots). Thus one past is is one past the end.
+		if (*it != UINT16_MAX) {
+			return ++it;
+		}
+		if (it == colors.begin()) { // Is it still safe to decrement?
+			return it; // No? Then the palette is empty.
+		}
+	}
```
Comment (Rangi42) [inline src/gfx/main.cpp:863]: I had to think a bit about the iteration logic here. I believe this would be simpler, and also valid even if it's somehow called on an empty palette:
```suggestion
	// Since the palette may contain gaps, we must scan from the end.
	for (auto it = colors.end(); it != colors.begin(); --it) {
		// If the iterator points after a non-empty element, then it is the end
		// (being followed only by empty slots). Thus one past it is one past the end.
		if (*(it - 1) != UINT16_MAX) {
			return it;
		}
	}
	return colors.end(); // The palette is empty.
```

File: man/rgbgfx.1 (hunk 2 above)
Comment (ISSOtm) [reply man/rgbgfx.1:134]: Agreed! Though I think it'd be better part of a separate PR, to avoid "inflating" the scope of this one.

File: src/gfx/main.cpp [hunk 6]
```diff
@@ -840,18 +847,42 @@ auto Palette::begin() -> decltype(colors)::iterator {
...
```
Comment (Rangi42) [inline src/gfx/main.cpp:883]: Whee, code duplication. :/ (Can we possibly have the `const` version delegate to the non-`const` one in C++?)
```suggestion
	// Since the palette may contain gaps, we must scan from the end.
	for (auto it = colors.end(); it != colors.begin(); --it) {
		// If the iterator points after a non-empty element, then it is the end
		// (being followed only by empty slots). Thus one past it is one past the end.
		if (*(it - 1) != UINT16_MAX) {
			return it;
		}
	}
	return colors.end(); // The palette is empty.
```
```

File: src/gfx/pal_spec.cpp [hunk 7]
```diff
@@ -88,25 +89,31 @@ void parseInlinePalSpec(char const * const rawArg) {
...
-			break;
-		case 0:
-			parseError(n - 1, 1, "Missing color after '#'");
-			return;
-		default:
-			parseError(n, pos - n, "Unknown color specification");
-			return;
+		std::optional<Rgba> &color = options.palSpec.back()[nbColors];
+		// Check for #none first.
+		if (arg.compare(n, 4, "none"sv) == 0 || arg.compare(n, 4, "NONE"sv) == 0) {
```
Comment (Rangi42) [inline src/gfx/pal_spec.cpp:94]: Looks like this would allow you to omit the comma after `none`: e.g. `-c "#FFF,#none#none#000"`.

File: src/gfx/main.cpp [hunk 8]
```diff
@@ -761,9 +761,16 @@ int main(int argc, char *argv[]) {
...
 		if (options.palSpecType == Options::EXPLICIT) {
 			fputs("\t[\n", stderr);
-			for (std::array<Rgba, 4> const &pal : options.palSpec) {
-				fprintf(stderr, "\t\t#%06x, #%06x, #%06x, #%06x,\n", pal[0].toCSS() >> 8,
-				        pal[1].toCSS() >> 8, pal[2].toCSS() >> 8, pal[3].toCSS() >> 8);
+			for (const auto &pal : options.palSpec) {
+				fputs("\t\t", stderr);
+				for (auto& color : pal) {
+					if (color) {
+						fprintf(stderr, "#%06x, ", color.value().toCSS() >> 8);
```
Comment (ISSOtm) [inline src/gfx/main.cpp:768]: A little conciseness nitpick.

```suggestion
						fprintf(stderr, "#%06x, ", color->toCSS() >> 8);
```

File: src/gfx/reverse.cpp [hunk 9]
```diff
@@ -171,13 +171,24 @@ void reverse() {
 			        options.nbPalettes);
 		}
 
-		if (options.palSpecType == Options::EXPLICIT && palettes != options.palSpec) {
-			warning("Colors in the palette file do not match those specified with `-c`!");
+		if (options.palSpecType == Options::EXPLICIT) {
+			for (auto& palette : palettes) {
+				for (auto& specPalette : options.palSpec) {
+					for (auto& color : palette) {
+						for (auto& specColor : specPalette) {
```
Comment (Rangi42) [inline src/gfx/reverse.cpp:178]: Formatting convention. (Also, lol, that's some deep looping.)
```suggestion
			for (auto &palette : palettes) {
				for (auto &specPalette : options.palSpec) {
					for (auto &color : palette) {
						for (auto &specColor : specPalette) {
```

File: src/gfx/process.cpp [hunk 10]
```diff
@@ -576,8 +576,11 @@ static std::tuple<DefaultInitVec<size_t>, std::vector<Palette>>
 	// Convert the palette spec to actual palettes
 	std::vector<Palette> palettes(options.palSpec.size());
 	for (auto [spec, pal] : zip(options.palSpec, palettes)) {
-		for (size_t i = 0; i < options.nbColorsPerPal && spec[i].isOpaque(); ++i) {
-			pal[i] = spec[i].cgbColor();
+		for (size_t i = 0; i < options.nbColorsPerPal && (!spec[i] || spec[i].value().isOpaque()); ++i) {
```
Comment (ISSOtm) [inline src/gfx/process.cpp:579]:
```suggestion
		for (size_t i = 0; i < options.nbColorsPerPal && (!spec[i] || spec[i]->isOpaque()); ++i) {
```

File: src/gfx/process.cpp [hunk 11]
```diff
@@ -576,8 +576,11 @@ static std::tuple<DefaultInitVec<size_t>, std::vector<Palette>>
...
+			// If the spec has a gap, there's no need to copy anything.
+			if (spec[i]) {
+				pal[i] = spec[i].value().cgbColor();
```
Comment (ISSOtm) [inline src/gfx/process.cpp:582]:
```suggestion
				pal[i] = spec[i]->cgbColor();
```

**Review by Rangi42 — CHANGES_REQUESTED** [review]

File: man/rgbgfx.1 [hunk 12]
```diff
@@ -125,11 +125,14 @@ begins with a hash character
...
 It should contain a comma-separated list of hexadecimal colors, each beginning with a hash.
-Colors in are accepted either as
+Colors are accepted either as
 .Ql #rgb
 or
 .Ql #rrggbb
 format.
+To leave a gap in the palette,
+.Ql #none
+can be used instead of one or more colors.
```
Comment (ISSOtm) [inline man/rgbgfx.1:135]:
```suggestion
To leave one or more gaps in the palette,
.Ql #none
can be used instead of any color.
```

I *think* this expresses the intent more clearly? `#none` can be used zero or more times, but each time substitutes for exactly one colour...

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/gfx/reverse.cpp (hunk 9 above)
Comment (ISSOtm) [reply src/gfx/reverse.cpp:178]: You should be able to call [`std::equal`](https://en.cppreference.com/w/cpp/algorithm/equal) with a custom comparator, then. (That's overload #3.)

File: src/gfx/reverse.cpp [hunk 13]
```diff
@@ -171,13 +171,24 @@ void reverse() {
...
+							if (!specColor || color != specColor.value()) {
```
Comment (ISSOtm) [inline src/gfx/reverse.cpp:179]:
```suggestion
							if (!specColor || color != *specColor) {
```

Comment (ISSOtm) [inline src/gfx/reverse.cpp:179]: Hmm, warning if `!specColor` means that any `-r N -c '...,#none,...'` will warn... shouldn't this be

```suggestion
							if (specColor && color != specColor.value()) {
```

?

File: src/gfx/reverse.cpp [hunk 14]
```diff
@@ -171,13 +171,24 @@ void reverse() {
...
+					}
+				}
+			}
 		}
 	} else if (options.palSpecType == Options::EMBEDDED) {
 		warning("An embedded palette was requested, but no palette file was specified; ignoring request.");
 	} else if (options.palSpecType == Options::EXPLICIT) {
-		palettes = std::move(options.palSpec); // We won't be using it again.
+		// TODO: what was this for?
+		//palettes = std::move(options.palSpec); // We won't be using it again.
```
Comment (ISSOtm) [inline src/gfx/reverse.cpp:191]: Why was this removed?

File: src/gfx/main.cpp (hunk 6 above)
Comment (ISSOtm) [reply src/gfx/main.cpp:883]: You **can** delegate, by using... `const_cast`. It's okay as long as `this` isn't modified, but then you're "unbuckling the safety belt" so to speak.

The alternative is using macros. Not sure we wanna go *there*...

File: src/gfx/pal_spec.cpp (hunk 7 above)
Comment (ISSOtm) [reply src/gfx/pal_spec.cpp:94]: Isn't the comma checked afterwards? The `else` branch doesn't check for it either.

File: src/gfx/main.cpp (hunk 6 above)
Comment (Rangi42) [reply src/gfx/main.cpp:883]: Seems like casting non-`const` to `const` when it's returned would be safe and concise.

File: src/gfx/main.cpp (hunk 4 above)
Comment (ISSOtm) [reply src/gfx/main.cpp:766]: `clang-format` should be run when this PR is ready to be merged.

File: src/gfx/main.cpp (hunk 5 above)
Comment (ISSOtm) [reply src/gfx/main.cpp:863]: Aside from the last line not being `return colors.begin();`, yes, it's correct.

File: src/gfx/main.cpp (hunk 6 above)
Comment (ISSOtm) [reply src/gfx/main.cpp:883]: Since the functions are close by and short, I'd rather live with that code duplicated.

**Comment by ISSOtm** [comment]
Test cases that should be added:

- [ ] Does `#none` correctly round-trip through `-r`?
- [ ] Can the comma be omitted after `#none`?

File: src/gfx/pal_spec.cpp (hunk 7 above)
Comment (eievui5) [reply src/gfx/pal_spec.cpp:94]: Yup, this code is correct
```
$ ./rgbgfx ~/nonetest.png -c#000,#none#none,#fff
error: Unexpected character, expected ',', ';', or end of argument
In inline palette spec: #000,#none#none,#fff
                                  ^
Conversion aborted after 1 error
```

File: src/gfx/reverse.cpp (hunk 9 above)
Comment (ISSOtm) [reply src/gfx/reverse.cpp]: > It used to be a really simple `palette == specPalette` but that doesn't work now that they're different types.

Are they? `clangd` tells me that both of them are `std::vector<std::array<std::optional<Rgba>, 4>>`.

File: src/gfx/reverse.cpp [hunk 15]
```diff
@@ -312,7 +322,7 @@ void reverse() {
 				uint8_t *ptr = &rowPtrs[y][tx * 8 * SIZEOF_PIXEL];
 				for (uint8_t x = 0; x < 8; ++x) {
 					uint8_t bit0 = bitplane0 & 0x80, bit1 = bitplane1 & 0x80;
-					Rgba const &pixel = palette[bit0 >> 7 | bit1 >> 6];
+					Rgba const &pixel = palette[bit0 >> 7 | bit1 >> 6].value();
```
Comment (ISSOtm) [inline src/gfx/reverse.cpp:325]:
```suggestion
					Rgba const &pixel = *palette[bit0 >> 7 | bit1 >> 6];
```

File: src/gfx/reverse.cpp (hunk 9 above)
Comment (eievui5) [reply src/gfx/reverse.cpp]: They were different types at the time of me writing that :P

File: src/gfx/reverse.cpp [hunk 16]
```diff
@@ -171,8 +171,18 @@ void reverse() {
...
+				for (auto &specPalette : options.palSpec) {
+					for (auto &color : palette) {
+						for (auto &specColor : specPalette) {
+							if (specColor && color != specColor.value()) {
+								warning("Colors in the palette file do not match those specified with `-c`!");
+							}
+						}
+					}
+				}
+			}
```
Comment (ISSOtm) [inline src/gfx/reverse.cpp:185]:
```suggestion
		if (options.palSpecType == Options::EXPLICIT && palettes != options.palSpec) {
			warning("Colors in the palette file do not match those specified with `-c`!");
```

Now that both are the same type, can they be compared directly again?

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/gfx/main.cpp (hunk 5 above)
Comment (Rangi42) [reply src/gfx/main.cpp:863]: Even in the general case, wouldn't `it - 1` be at least as cheap as `--it` followed by `++it` undoing it?

File: src/gfx/main.cpp (hunk 6 above)
Comment (Rangi42) [reply src/gfx/main.cpp:883]: Okay, then this diff would need applying too to keep the `iterator` and `const_iterator` in sync.

File: src/gfx/main.cpp (hunk 5 above)
Comment (ISSOtm) [reply src/gfx/main.cpp:863]: No, because it has to create a temporary, i.e. make a copy.

File: src/gfx/main.cpp (hunk 6 above)
Comment (ISSOtm) [reply src/gfx/main.cpp:883]: By the way, `*(it - 1)` can be written as `it[-1]`. It can even be a little faster for non-pointer iterators, since they can return the value instead of having to create a temporary iter :)

File: src/gfx/main.cpp (hunk 5 above)
Comment (Rangi42) [reply src/gfx/main.cpp:863]: If only something like` std::string`'s`find_last_not_of` could apply here.

Comment (ISSOtm) [reply src/gfx/main.cpp:863]: Oh, I guess `rbegin()` and `rend()` could be used with the usual `std::find` to scan the array backwards. :man_facepalming:

Comment (Rangi42) [reply src/gfx/main.cpp:863]: Would we need to manually implement `rbegin`/`rend`/a `reverse_iterator`?

Comment (ISSOtm) [reply src/gfx/main.cpp:863]: No, but we can use them to quickly find the index of the last non-`UINT16_MAX` element <var>n</var>, and then `return colors.begin() + n;`.

Comment (Rangi42) [reply src/gfx/main.cpp:863]: Wait, actually, `colors` is a `std::array` which *is* reverse-iterable. So I think this is correct:
```c++
return (std::find_if(colors.rbegin(), colors.rend(), [](uint16_t c) { return c != UINT16_MAX; }) + 1).base();
```
(Note the avoidance of an off-by-one error with `.base()`.)

File: test/gfx/rgbgfx_test.cpp [hunk 17]
```diff
@@ -281,6 +281,19 @@ class Png {
 };
 
 static char *execProg(char const *name, char * const *argv) {
+	auto formatArgv = [&argv] {
+		static std::string buf; // It's fine to use a static string:
```
Comment (Rangi42) [inline test/gfx/rgbgfx_test.cpp:285]: It may be "fine" (harmless), but is it for a particular reason? Is it just to avoid allocating a string every time `execProg` is called? If so, IMO just don't make it `static`, and then no need to `clear` either.

File: src/gfx/reverse.cpp (hunk 16 above)
Comment (Rangi42) [reply src/gfx/reverse.cpp:185]: Apparently yes.

File: test/gfx/rgbgfx_test.cpp (hunk 17 above)
Comment (ISSOtm) [reply test/gfx/rgbgfx_test.cpp:285]: No, it's so that the returned `c_str()` lives long enough. I'm not sure whether returning a `std::string` and then using *that*'s `.c_str()` is safe, or a UAF; so I chose something which I knew should work.

If making this return the `std::string` directly would work correctly, then by all means it should be done.

**Review by ISSOtm — DISMISSED** [review]
The code looks all good to me! :tada: 

That just leaves the test cases mentioned above left to add.

File: test/gfx/rgbgfx_test.cpp (hunk 17 above)
Comment (Rangi42) [reply test/gfx/rgbgfx_test.cpp:285]: Okay, thanks; then a comment to that effect would be helpful.

**Comment by ISSOtm** [comment]
Ping @eievui5?

**Review by ISSOtm — APPROVED** [review]
Thank you!

**Review by Rangi42 — COMMENTED** [review]
The new tests look good! You're offline atm so I hope you don't mind me pushing a couple tweaks to it again.

File: test/gfx/rgbgfx_test.cpp (hunk 17 above)
Comment (Rangi42) [reply test/gfx/rgbgfx_test.cpp:285]: (Ooh, just noticed that the no-arg anonymous function can be written as `[] {...}`, with no `()`. Neat!)

**Review by Rangi42 — APPROVED** [review]
Thank you!
